packages/platforms/delta/arm64/tx4810-dn/modules/builds/src/arm64-delta-tx4810-dn-led.c

```diff
@@ -0,0 +1,391 @@
+    LED_TYPE_FAN,
+    LED_TYPE_SYS,
+    NUM_OF_LED
+};
+
+enum led_light_mode {
+    LED_MODE_OFF = 0,
+    LED_MODE_GREEN,
+    LED_MODE_AMBER,
+    LED_MODE_UNKNOWN,
+};
+
+struct tx4810_dn_gpio {
+    const char *name;
+    unsigned int gpio;
+};
+
+static struct tx4810_dn_gpio tx4810_dn_led_gpio[] = {
+    {"LED PSU Green", 35},
+    {"LED PSU Amber", 34},
+    {"LED FAN Green", 46},
+    {"LED FAN Amber", 43},
+    {"LED SYS Green", 63},
+    {"LED SYS Amber", 66},
+};
+
+#define TX4810_DN_COMMON_LED_MAP  (BIT(LED_TYPE_PSU) | BIT(LED_TYPE_FAN) | \
+                                 BIT(LED_TYPE_SYS))
+
+struct tx4810_dn_led_data {
+    struct platform_device *pdev;
+    struct mutex     update_lock;
+    char             valid;             /* != 0 if registers are valid */
+    unsigned long    last_updated;      /* In jiffies */
+    int              led_map;
+    u8               led_gpio_val[6];
+};
+
+static struct tx4810_dn_led_data  *ledctl = NULL;
+
+
+static enum led_light_mode gpio_val_to_led_mode(int led_green, int led_amber)
+{
+    if (led_green && !led_amber)
+        return LED_MODE_GREEN;
+    else if (!led_green && led_amber)
+        return LED_MODE_AMBER;
+    else
+        return LED_MODE_OFF;
+}
+
+static void led_mode_to_gpio_value(enum led_light_mode mode,
+                                               int *led_green, int *led_amber)
+{
+    switch (mode)
+    {
+    case LED_MODE_GREEN:
+        *led_green = 1;
+        *led_amber = 0;
+        break;
+    case LED_MODE_AMBER:
+        *led_green = 0;
+        *led_amber = 1;
+        break;
+    default:
+        *led_green = 0;
+        *led_amber = 0;
+        break;
+    }
+}
+
+static int tx4810_dn_gpio_map(struct platform_device *pdev)
+{
+    int i, err;
+
+    for (i = 0; i < ARRAY_SIZE(tx4810_dn_led_gpio); i++) {
+        err = gpio_request(tx4810_dn_led_gpio[i].gpio, tx4810_dn_led_gpio[i].name);
+        if (err) {
+            dev_err(&pdev->dev, "error mapping gpio %s: %d\n",
+                    tx4810_dn_led_gpio[i].name, err);
+            goto err_request;
+        }
+        gpio_direction_output(tx4810_dn_led_gpio[i].gpio, 0);
+    }
+
+    return 0;
+
+err_request:
+    while (--i >= 0)
+        gpio_free(tx4810_dn_led_gpio[i].gpio);
+    return err;
+}
+
+static void tx4810_dn_gpio_unmap(void)
+{
+    int i;
+
+    for (i = 0; i < ARRAY_SIZE(tx4810_dn_led_gpio); i++)
+        gpio_free(tx4810_dn_led_gpio[i].gpio);
+}
+
+static int tx4810_dn_led_read_value(unsigned int gpio)
+{
+    return gpio_get_value(gpio);
+}
+
+static void tx4810_dn_led_write_value(unsigned int gpio, u8 value)
+{
+    return gpio_set_value(gpio, value);
+}
+
+static void tx4810_dn_led_update(void)
+{
+    mutex_lock(&ledctl->update_lock);
+
+    if (time_after(jiffies, ledctl->last_updated + HZ + HZ / 2)
+        || !ledctl->valid) {
+        int i;
+
+        dev_dbg(&ledctl->pdev->dev, "Starting tx4810_dn_led update\n");
+
+        /* Update LED data
+         */
```

Comment:
Should be one line?

Reply:
The above suggestions are all taken into led and cpld driver. Hope it is not too late. Thanks a lot.